### ISSUE TITLE
Crossfade

### ISSUE THREAD

**Issue by GhoulBoii** [KRTirtho/spotube#783]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Pain

Spotify has crossfading between songs which easily transitions from one song to another. Spotube, on the other hand, abruptly stops for a couple seconds before playing the next song. If crossfading from one song to another could be added, that would be appreciated.

### Suggested solution

Pre-caching a couple of seconds of the next song in the queue and then having a user-configurable crossfade time option to crossfade to the next song.

### Useful resources

https://github.com/z-huang/InnerTune/issues/889

**Comment by jojommeke** [comment]
i think this is a good thing can anyone do something for this feature?

**Comment by rramirezg22** [comment]
yes! that couple of seconds are really annoying. Hope you could add this feature.

**Comment by JustAdreamerFL** [comment]
That sounds difficult but I would really appreciate this feature.

**Comment by javidiez** [comment]
Any update on this?

**Comment by tectrixdev** [comment]
(Still relevant.)

**Comment by henikkataaaria** [comment]
Hey everyone, I'm a heavy music listener using Spotube, and I've been tracking the gapless playback discussions. I know the app recently migrated to media_kit and that background preloading for the next track's URL is already implemented around the 20% mark.

However, it seems the 1-2 second gap still occurs because when Track A ends, the player invokes a manual handoff (likely re-initializing the audio decoders via a fresh open call).

To bypass this engine reset, could we leverage media_kit’s native Playlist class and append the preloaded URL directly to the active native queue via player.add() before Track A finishes? That way, the underlying C++ audio engine (MPV/ExoPlayer) can handle the stream transition seamlessly without a decoder flush.

Here is a conceptual blueprint of what that queue logic might look like in the audio service:
Dart

// Conceptual look-ahead append to media_kit playlist
Future<void> onPreloadNextTrack(Track spotubeTrackB) async {
  // Utilize existing logic to resolve the stream URL
  final String resolvedStreamUrlB = await extractStreamUrl(spotubeTrackB);
  
  // Append directly to the active media_kit playlist object
  // This allows the native C++ engine to handle gapless buffering
  await player.add(Media(resolvedStreamUrlB));
}

Thanks to KRTirtho and all the contributors for the amazing work on the media_kit migration so far!
